charms/worker/k8s/src/charm.py

```diff
@@ -1061,14 +1120,102 @@ def _get_external_kubeconfig(self, event: ops.ActionEvent):
         try:
             server = event.params.get("server")
             if not server:
-                log.info("No server requested, use public-address")
-                server = f"{_get_public_address()}:6443"
+                log.info("No server requested, use public address")
+
+                server = self._get_public_address()
+                if not server:
+                    event.fail("Failed to get public address. Check logs for details.")
+                    return
+
+                port = (
+                    str(EXTERNAL_LOAD_BALANCER_PORT)
+                    if self.external_load_balancer.is_available
+                    else str(APISERVER_PORT)
+                )
+
+                server = build_url(server, port, "https")
+                log.info("Formatted server address: %s", server)
             log.info("Requesting kubeconfig for server=%s", server)
             resp = self.api_manager.get_kubeconfig(server)
             event.set_results({"kubeconfig": resp})
         except (InvalidResponseError, K8sdConnectionError) as e:
             event.fail(f"Failed to retrieve kubeconfig: {e}")
 
+    def _get_public_address(self) -> Optional[str]:
+        """Get the most public address either from external load balancer or from juju.
+
+        If the external load balancer is available and the unit is a control-plane unit,
+        the external load balancer address will be used. Otherwise, the juju public address
+        will be used.
+        NOTE: Don't ignore the unit's IP in the extra SANs just because there's a load balancer.
+
+        Returns:
+            str: public ip address of the unit
+            None: if the public address is not available
+        """
+        if self.is_control_plane and self.external_load_balancer.is_available:
+            log.info("Using external load balancer address as the public address")
+            return self._get_external_load_balancer_address()
+        log.info("Using juju public address as the public address")
+        return _get_juju_public_address()
+
+    def _get_external_load_balancer_address(self) -> Optional[str]:
+        """Get the external load balancer address.
+
+        Returns:
+            None: If the external load balancer address is not available.
+            str: The external load balancer address.
+        """
+        if not self.is_control_plane:
+            log.error("External load balancer address is only available for control-plane units")
+            return None
+
+        response = self.external_load_balancer.get_response(EXTERNAL_LOAD_BALANCER_RESPONSE_NAME)
+
+        if not response:
+            log.error("No response from external load balancer when trying to get address")
+            return None
```

Comment:
This might be a transient error. What if we adjust the log level here?
```suggestion
        if not response:
            log.info("Response from external load balancer is not yet available.")
            return None
```